ftl/common/cache_runner.py

```diff
@@ -0,0 +1,232 @@
+                self._remove_entry(entry)
+
+    def _remove_entry(self, entry):
+        # delete entry from mapping and cache
+        del self._mappings[entry]
+        docker_session.Delete(self._tag(entry), self._creds, self._transport)
+
+    def populate_cache_entries(self, existing_entries):
+        # for each package, either verify it is already in the cache,
+        # or build the image and push it to the cache
+        for package in self._packages:
+            if package:
+                try:
+                    name = None
+                    version = None
+                    if self._language == PHP:
+                        name, version = package.split(':')
+                    elif self._language == PYTHON:
+                        name, version = package.split('==')
+                    if name not in existing_entries:
+                        # builder._pip_install() expects the double equals
+                        # on the version
+                        self._build_image_and_push(name, '==' + version)
+                except ValueError:
+                    logging.error('Encountered malformed package: {0}'.format(
+                        package))
+
+    def _build_image_and_push(self, package_name, package_version):
+        logging.info('building package {name}, version {version}'.format(
+            name=package_name,
+            version=package_version))
+        image = None
+        builder = None
+        if self._language == PHP:
+            builder = php_builder.PhaseTwoLayerBuilder(
+                pkg_descriptor=(package_name, package_version))
+        elif self._language == PYTHON:
+            interpreter_builder = python_builder.InterpreterLayerBuilder()
+            interpreter_builder.BuildLayer()
+            builder = python_builder.PipfileLayerBuilder(
+                pkg_descriptor=(package_name, package_version),
+                wheel_dir=self._setup_pip_and_wheel(),
+                dep_img_lyr=interpreter_builder)
+        if not builder:
+            logging.error('Could not find builder for language {0}'.format(
+                self._language))
+            sys.exit(1)
+        builder.BuildLayer()
+        # since we only have one layer, just use the builder's image
+        # itself as the final image
+        image = builder._img
+
+        # TODO(nkubala): we should refactor AppendLayersIntoImage to not
+        # have to set a base image
+        # image = ftl_util.AppendLayersIntoImage([builder])
+
+        key = builder.GetCacheKey()
+        tag = self._tag(key)
+
+        with docker_session.Push(tag, self._creds,
+                                 self._transport, threads=2) as session:
+            session.upload(image)
+        self._mappings['%s:%s' % (package_name, package_version)] = key
+
+    def write_mapping_to_workspace(self):
+        with open(MAPPING_FILE.format(language=self._language), 'w') as f:
+            f.write(json.dumps(self._mappings))
```

Comment:
not: json.dump(self._mappings, f) is the same as this.

Reply:
nice